data_parallel_tensor.py

```diff
@@ -0,0 +1,396 @@
+import concurrent.futures as futures
+
+_ = torch.manual_seed(0)
+aten = torch.ops.aten
+NUM_DEVICES = 8
+PARALLEL_DISPATCH = False
+ALL_REDUCE = True
+
+
+class DPTensorType(Enum):
+    replicated = auto()  # This tensor will be replicated across all the devices
+    distributed_batch = (
+        auto()
+    )  # This tensor will be sharded along the first/batch dimension across
+    # the devices, NOTE: only equal chunk sizes are supported
+    distributed = (
+        auto()
+    )  # This is a list of tensors, each of which rests on different devices
+
+
+class DataParallelTensor(torch.Tensor):
+    # This class is a tensor subclass that stores a list of tensors with the aim
+    # DataParallelTensors(DPT) are categorized in three ways
+    # 1) replicated: When a single tensor is supplied, it is replicated across
+    #   all the devices by using broadcast
+    # 2) distributed: DPT can also be initialized by supplying a list/tuple of tensors
+    #   if the elements rest on different devices, they will just be wrapped in DPT
+    #   else the elements are scattered to different devices
+    # 3) distributed batch: This type of DPT tensor is created by sharding the input tensor across
+    #   a specified sharding dimension (default: 0). Currently only equal chunk sizes are supported.
+
+    elem: List[torch.Tensor]
+
+    if torch.cuda.is_available():
+        # device_ids: List[int] = _get_all_device_indices()
+        device_ids = [i for i in range(NUM_DEVICES)]
+        if PARALLEL_DISPATCH:
+            num_threads: int = len(device_ids)
+            threadpool: futures.ThreadPoolExecutor = futures.ThreadPoolExecutor(
+                max_workers=num_threads
+            )
+    __slots__ = ["elem"]
+
+    @staticmethod
+    def __new__(
+        cls,
+        elem: Union[torch.Tensor, List[torch.Tensor], Tuple[torch.Tensor]],
+        func: Optional[Any] = None,
+        dpt_type: DPTensorType = DPTensorType.replicated,
+        batch_dim: Optional[int] = 0,
+    ):
+
+        if dpt_type == DPTensorType.replicated:
+            # NOTE: If the input is None, we return None
+            if elem is None:
+                return None
+            assert isinstance(elem, torch.Tensor)
+            # NOTE: For handling meta tensors, if the device of an input tensor is meta,
+            # we just return the first element in such a list/tuple
+            if elem.device == device("meta"):
+                return elem
+
+            with torch.no_grad():
+                dpt: List[torch.Tensor] = comm.broadcast(elem, devices=cls.device_ids)
+
+        elif dpt_type == DPTensorType.distributed:
+            assert isinstance(elem, list) or isinstance(elem, tuple)
+            # NOTE: For handling None values, if any element of a list/tuple is None, we return None
+            check_none = [True if e is None else False for e in elem]
+            if any(check_none):
+                return None
```

Comment:
What do you think of adding an assertion - if one is None, all should be None?

Reply:
This is a design decision. We could potentially have a DPT that supports Tensors on some devices and None for some. Need to think more on this. Adding an assertion might not be the right thing to do since it is not always guaranteed that all the values should be None. We should keep this comment open for now.

Reply:
The reason we just decided to return None was since some operators make boolean checks on input tensors to see if they are none and since we have a single meta-data for dpt, None values become a concern.

Reply:
I see. In what scenarios, do we have None on some devices, and tensors on others? Have you seen any such scenarios in your testcase?

Reply:
So according to our discussion, I make a single check to see if the values in a list are tensors or not, if they are not tensors then we return only a single value if all of them are same else raise an error. We need to insert these checks at two places:
1. DPT Constructor: When a list/tuple of values are supplied
2. When we are wrapping the output list 
Thanks for the suggestion code looks a lot cleaner and robust now :)